File: Module A/performanceTests/stack_vs_heap.jl
using BenchmarkTools
using StaticArrays
using Printf

const N = 1000000

function mysum(vector, samples, reps)
  s = 0
  for i = 1:reps
    s -= s/i
    for j = 1:samples
      s += vector[j]
    end
  end
  return s
end

function (@main)(args)
  samples = ( length(args) >= 1 ) ? parse(Int, args[1]) : N;
  s = 0;
  r = ( length(args) >= 2 ) ? parse(Int, args[2]) : 1000;

  # Init Data
  heapArray  = rand(Int, samples)
  stackArray = MVector{samples,Int}(undef)
  for i in 1:samples
    stackArray[i] = heapArray[i]
  end

  @printf("Test with %d samples:\n", samples);

  # Heap Allocation
  println("Heap allocation")
  @time begin
    s = mysum(heapArray, samples, r);
  end
  println("result: ", s)

  # Stack Allocation
  println("Stack allocation")
  @time begin
    s = mysum(stackArray, samples, r);
  end
  println("result: ", s)
end

File: Module A/performanceTests/stack_vs_heap.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <alloca.h>

#define N 1000000

void sum(int* vector, volatile int* s, int samples, int reps) {
  *s = 0;
  for(volatile int i=0; i<reps; i++) {
    *s -= *s/i;
    for(volatile int j=0; j<samples; j++)
      *s += vector[j];
  }
  return;
}

int main(int argc, char** argv) {
  int* heapArray =  (int*)malloc( sizeof(int) * N );
  int* stackArray = (int*)malloc( sizeof(int) * N );

  volatile int samples = ( argc > 1 ) ? atoi(argv[1]) : N;
  volatile int s = 0;
  volatile int r = ( argc > 2 ) ? atoi(argv[2]) : 10;

  clock_t start, end;
  double used_time;

  // Init Data
  srand( time(NULL) );
  for(int i = 0; i < samples; i++) {
    heapArray[i] = stackArray[i] = rand();
  }

  printf("Test with %d samples:\n", samples);

  // Heap Allocation
  start = clock();
    sum(heapArray, &s, samples, r);
  end = clock();
  used_time = ((double)(end-start)) / CLOCKS_PER_SEC;
  printf("Heap allocation:\n\tresult: %d\n\tCPU time: %lf\n",s,used_time);

  // Stack Allocation
  start = clock();
    sum(stackArray, &s, samples, r);
  end = clock();
  used_time = ((double)(end-start)) / CLOCKS_PER_SEC;
  printf("Stack allocation:\n\tresult: %d\n\tCPU time: %lf\n",s,used_time);

  free(heapArray);
  return 0;
}
